plugins/core/gov.nasa.jpl.imce.oml.viewpoint/src/gov/nasa/jpl/imce/oml/viewpoint/TerminologyService.xtend

```diff
@@ -0,0 +1,337 @@
+				// Put Aspects on queue to travel
+				queue.add(entity)
+			}			
+		]
+		visited.add(c)		
+		
+		// Search graph for connected concepts along path
+		// of root concept passed as method arg
+		while(queue.peek() != null){
+			val node = queue.poll()
+			// Check for entity relationships to Concepts
+			var foundConnnection = false
+			for(e : graph.get(node)){
+				// node <relOrAx> n
+				val n = e.key
+				val relOrAx = e.value
+				if(((relOrAx instanceof EntityRelationship) || (relOrAx instanceof EntityRestrictionAxiom)) && 
+					(n instanceof Concept)
+				){
+					    relationships.add(relOrAx)
+					    visited.add(n)
+					    // No need to go further. 
+					    // Don't travel further along this path
+					    foundConnnection = true
+			    } else {
+			    	if(foundConnnection){
+			    		// Already found connection
+			    		visited.add(n)
+			    	}
+			    	else if(!((relOrAx instanceof EntityRelationship) || (relOrAx instanceof EntityRestrictionAxiom)) && 
+			    		(n instanceof Concept)){
+			    		// Axiom to Concept
+					    visited.add(n)
+			        } else {
+			    	      // n is Aspect and relOrAx is an Axiom AND connection not found yet
+			    	      // put n on queue to search that path	
+			    	      if(!visited.contains(n)){
+			    	          visited.add(n)
+			    	          queue.add(n)
+			    	      }		
+			        }    	    			    	
+			    }
+		    }
+		}
+		
+		relationships
+	}
+	
+	/*
+	 * Gets all {@link ReifiedRelationship}s that are indirectly connected
+	 * to the {@link Concept} 'c' in which 'c' would be the source
+	 * @param c Root {@link Concept}
+	 * @return Set<ReifiedRelationship> Set of indirectly connected relationships
+	 */
+	def Set<ReifiedRelationship> getIndirectRelationshipsWithRootAsSource(Concept c){
+		val relationships = new HashSet<ReifiedRelationship>
+		
+        getUsageReltionships(c).
+		filter(ReifiedRelationship).
+		filter[f | f.source instanceof Aspect].
+		forEach[t | 
+			relationships.add(t as ReifiedRelationship)
+		]
+		
+		relationships		
+	}	
+	
+	/*
+	 * Gets all {@link ReifiedRelationship}s that are indirectly connected
+	 * to the {@link Concept} 'c' in which 'c' would be the target
+	 * @param c Root {@link Concept}
+	 * @return Set<ReifiedRelationship> Set of indirectly connected relationships
+	 */
+	def Set<ReifiedRelationship> getIndirectRelationshipsWithRootAsTarget(Concept c){
+	    val relationships = new HashSet<ReifiedRelationship>
+		
+        getUsageReltionships(c).
+		filter(ReifiedRelationship).
+		filter[f | f.target instanceof Aspect].
+		forEach[t | 
+			relationships.add(t as ReifiedRelationship)
+		]
+		
+		relationships			
+	}
+	
+	private def Map<Entity,List<Entry<Entity,TerminologyBoxStatement>>> buildEntityGraph(Entity c){
+				
+		// Key: Entity
+		// Value: List of Entity, Relationship/Axiom pairs
+		val graph = new HashMap<Entity,List<Entry<Entity,TerminologyBoxStatement>>>
+		
+		// Build graph
+		c.tbox.boxStatements.

```

Comment:
Why limit the graph to the tbox where the entity is defined?

If you add:

```
@Inject extension OMLScopeExtensions
```

Then you could use:

```
c.tbox.allEntitiesScope
```

to get all entities that are directly or indirectly defined in the entity's tbox or an extended tbox.